Forge/src/main/java/me/ferdz/placeableitems/client/model/FluidModel.java

```diff
@@ -0,0 +1,133 @@
+package me.ferdz.placeableitems.client.model;
+
+import java.util.IdentityHashMap;
+import java.util.Map;
+
+import me.ferdz.placeableitems.block.component.impl.FluidHolderBlockComponent;
+import me.ferdz.placeableitems.client.AllVertexBoundingBox;
+import me.ferdz.placeableitems.client.renderer.tileentity.FluidHolderRenderer;
+
+import net.minecraft.block.BlockState;
+import net.minecraft.client.renderer.Vector3d;
+import net.minecraft.util.Direction;
+
+import net.minecraftforge.fml.event.lifecycle.FMLClientSetupEvent;
+
+/**
+ * Represents a model definition for a fluid renderable in a {@link FluidHolderBlockComponent}.
+ * Models should cache model data where possible (including rotation points and render bounds)
+ * and return them where necessary.
+ * <p>
+ * All states will have their render bounds cached and may be
+ * recalculated at any time such that {@link #shouldRecalculate(BlockState)} returns true. Please
+ * be cautious when returning true in the aforementioned method such that any computationally-expensive
+ * calculations will be re-run (however {@link #calculateRenderBounds(BlockState)} is implemented).
+ * <p>
+ * FluidModel implementations should be registered in a {@link FMLClientSetupEvent} with the
+ * {@link FluidHolderRenderer#bind(me.ferdz.placeableitems.block.PlaceableItemsBlock, FluidModel)}
+ * method. If a {@link FluidHolderBlockComponent#shouldRenderFluid()} returns true but has no registered
+ * FluidModel, an exception will be thrown an the client will crash.
+ *
+ * @see FluidHolderRenderer#bind(me.ferdz.placeableitems.block.PlaceableItemsBlock, FluidModel)
+ *
+ * @author Parker Hawke - Choco
+ */
+public abstract class FluidModel {
+
+    private static final Vector3d ZEROED_OFFSET = new Vector3d(0.0D, 0.0D, 0.0D);
+
+    private final Map<BlockState, AllVertexBoundingBox> renderCache = new IdentityHashMap<>();
+
+    /**
+     * Get the rendering bounds for this model from the render cache (if present) and optionally
+     * recalculate the cached model. If a model is not present, it will be calculated and cached.
+     *
+     * @param state the state whose model to retrieve
+     * @param recalculate whether or not to recalculate the model
+     *
+     * @return the rendering bounds
+     */
+    public final AllVertexBoundingBox getRenderBounds(BlockState state, boolean recalculate) {
```

Comment:
I feel like the `recalculate`  parameter doesn't belong here. Wouldn't it be better if it called `shouldRecalculate` directly? This would avoid having to do it in the renderer, where it shouldn't be aware of updates

Reply:
This is something I actually considered. However, I believe that flexibility should be given to the callers rather than having that value hard-coded. If you believe otherwise, I'm open to change this.

Reply:
I'd rather have it here, it'll reduce the complexity of using the model